status/status.go

```diff
@@ -37,6 +39,146 @@ import (
 // NamePrefix is a common name prefix for this service.
 const NamePrefix = "Red Hat Trusted App Test"
 
+// ScenarioReportStatus keep report status of git provider for the particular scenario
+type ScenarioReportStatus struct {
+	LastUpdateTime *time.Time `json:"lastUpdateTime"`
+}
+
+// SnapshotReportStatus keep report status of git provider for the snapshot
+type SnapshotReportStatus struct {
+	Scenarios map[string]*ScenarioReportStatus `json:"scenarios"`
+	dirty     bool
+}
+
+// SetLastUpdateTime updates the last udpate time of the given scenario to the given time
+func (srs *SnapshotReportStatus) SetLastUpdateTime(scenarioName string, t time.Time) {
+	srs.dirty = true
+	if scenario, ok := srs.Scenarios[scenarioName]; ok {
+		scenario.LastUpdateTime = &t
+		return
+	}
+
+	srs.Scenarios[scenarioName] = &ScenarioReportStatus{
+		LastUpdateTime: &t,
+	}
+}
+
+// IsNewer returns true if given scenario has newer time than the last updated
+func (srs *SnapshotReportStatus) IsNewer(scenarioName string, t time.Time) bool {
+	if scenario, ok := srs.Scenarios[scenarioName]; ok {
+		return scenario.LastUpdateTime.Before(t)
+	}
+
+	// no record, it's new
+	return true
+}
+
+// ToAnnotationString exports data in format for annotation
+func (srs *SnapshotReportStatus) ToAnnotationString() (string, error) {
+	byteVar, err := json.Marshal(srs)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal report status metadata: %w", err)
+	}
+	return string(byteVar), nil
+}
+
+// IsDirty returns true if there are new changes to be written
+func (srs *SnapshotReportStatus) IsDirty() bool {
+	return srs.dirty
+}
+
+// ResetDirty marks changes as synced to snapshot
+func (srs *SnapshotReportStatus) ResetDirty() {
+	srs.dirty = false
+}
+
+// NewSnapshotReportStatus creates new object
+func NewSnapshotReportStatus(jsondata string) (*SnapshotReportStatus, error) {
+	srs := SnapshotReportStatus{
+		Scenarios: map[string]*ScenarioReportStatus{},
+	}
+	if jsondata != "" {
+		err := json.Unmarshal([]byte(jsondata), &srs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal json: %w", err)
+		}
+	}
+	return &srs, nil
+}
+
+// NewSnapshotReportStatusFromSnapshot creates new SnapshotTestStatus struct from snapshot annotation
+func NewSnapshotReportStatusFromSnapshot(s *applicationapiv1alpha1.Snapshot) (*SnapshotReportStatus, error) {
+	annotations := map[string]string{}
```

Comment:
If we change L111-L118 to code as below, does it work?
```
if statusAnnotation, found := s.ObjectMeta.Annotations[gitops.SnapshotStatusReportAnnotation]; !found {
        statusAnnotation = ""
}
```

Reply:
if Annotations are nil, it would fail

Reply:
I try this and it works, but if it doesn't work for this PR, please ignore it. 
```
testSnapshot := &applicationapiv1alpha1.Snapshot{
	ObjectMeta: metav1.ObjectMeta{
		Name:      "snapshot-test",
		Namespace: "default",
		Labels: map[string]string{
			gitops.SnapshotTypeLabel:            "component",
			gitops.SnapshotComponentLabel:       hasComp.Name,
		},
	},
	Spec: applicationapiv1alpha1.SnapshotSpec{
		Application: hasApp.Name,
		Components: []applicationapiv1alpha1.SnapshotComponent{
			{
				Name:           hasComp.Name,
				ContainerImage: SampleImage,
			},
		},
	},
}
statusAnnotation, found := testSnapshot.ObjectMeta.Annotations["abc"]
if !found {
	statusAnnotation = "test"
}
```

Reply:
updated